test_apps/smokeTests/gridTools.cpp

```diff
@@ -0,0 +1,594 @@
+    float zIncrement = axis==Z ? (maxVal - minVal) / (dims[Z]-1) : 0;
+
+    float value = minVal;
+    for ( size_t k=0; k<z; k++ ) {
+        value = axis==Y ? minVal : value;      // reset value if we're ramping on Y
+        for ( size_t j=0; j<y; j++ ) {
+            value = axis==X ? minVal : value;  // reset value if we're ramping on X
+            for ( size_t i=0; i<x; i++ ) {
+                grid->SetValueIJK( i, j, k, value );
+                value += xIncrement;
+            }
+            value += yIncrement;
+        }
+        value += zIncrement;
+    }
+}
+
+// This function iterates across all nodes in a grid; making comparisons between the
+// data values returned by the functions GetValueAtIndex() and GetValue().
+bool CompareIndexToCoords( 
+    VAPoR::Grid* grid, 
+    double &rms,                // Root Mean Square error
+    size_t &numMissingValues,   // Counter for receiving MissingValue upon query
+    size_t &disagreements      // Counter for when AccessIJK() and GetValue() disagree
+) {
+    rms   = 0.f;
+    disagreements  = 0;
+    numMissingValues = 0;
+
+    std::vector< size_t > dims = grid->GetDimensions();
+    size_t x = dims[X];
+    size_t y = dims[Y];
+    size_t z = 1;
+    if ( dims.size() == 3 ) 
+        z = dims[Z];
+ 
+    double peak  = 0.f;
+    double sum     = 0;
+    for ( size_t k=0; k<z; k++ ) {
+        for ( size_t j=0; j<y; j++ ) {
+            for ( size_t i=0; i<x; i++ ) {
+                size_t indices[3] = { i, j, k };
+                //float trueValue = grid.AccessIJK( i, j, k );
+                double trueValue = grid->GetValueAtIndex( {i,j,k} );
+                
+                double coords[3];
+                grid->GetUserCoordinates( indices, coords );
+                float sampleValue = grid->GetValue( coords );
+                
+                if (sampleValue == grid->GetMissingValue()) {
+                    numMissingValues++;
+                    continue;
+                }
+
+                double error = abs( sampleValue - trueValue );
+
+                if (error != 0) {
+                    disagreements++;
+                }
+
+                if ( error > peak ) 
+                    peak = error;
+                sum += error*error;
+
+            }
+        }
+    }
+
+    rms = sqrt( sum / (x*y*z) );
+
+    if ( rms != 0 || disagreements > 0 )
+        return 1;
+    return 0;
+}
+
+bool isNotEqual(
+    double x, 
+    double y
+) {
+  const double epsilon = 1e-5;
+  return std::abs(x - y) > epsilon * std::abs(x);
+}
+
+size_t TestConstNodeIterator( 
+    const Grid *g, 
+    size_t& count, 
+    size_t& disagreements, 
+    double& time 
+) {
+    Grid::ConstNodeIterator itr;
+    Grid::ConstNodeIterator enditr = g->ConstNodeEnd();
+
+    double t0 = Wasp::GetTime();
+    itr = g->ConstNodeBegin();
+    
+    std::vector< size_t > dims = g->GetDimensions();
+
+    for ( ; itr!=enditr; ++itr) {
+        size_t i = count % dims[X];
```

Comment:
The utility function Wasp::VectorizeCoords will turn an offset into component coordinates and will not break if the grid dimension is not 2 or 3 as this code assumes.

Reply:
Good to know.  Fixed.